### PR TITLE
[ARCTIC-1025][FLINK] Fix data duplication even if there is a primary key table with upsert enabled

### LINKED ISSUE #1025: [Bug]: The same _file_offset in the Change Store causes data duplication even if there is a primary key table with upsert enabled
### What happened?

For tables with a primary key and the property 'write.upsert.enabled' set, when the same transaction id commits multiple changes to a primary key data, it can lead to data duplication. And the '_file_offset' of these duplicate data must be duplicated.
image

### Affects Versions

0.3.x

### What engines are you seeing the problem on?

Flink

### How to reproduce

1.Create an Arctic table with a primary key and start a Flink ingestion job
2.The same primary key is used for Insert and Update operations during a single transaction commit

### Relevant log output

_No response_

### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by zstraw**
FIX #1025 

## Why are the changes needed?
To make sure only one primary key in key table with upsert enabled, it's necessary to delete data firstly when update_after not followed by update_before.

It may lead to some unnecessary delete data if there are some data interspersed with other primary key data. e.g. K1-UB, K2-UB, K1-UA, K2-UB

## Brief change log

  - *The flink 1.12 and 1.14, 1.15 versions are affected.*

## How was this patch tested?
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink/v1.12/flink/src/test/java/com/netease/arctic/flink/table/TestKeyed.java [hunk 1]
```diff
@@ -441,12 +441,14 @@ public void testPartitionSinkFile() throws IOException {
   @Test
   public void testFileUpsert() throws IOException {
```
Comment (YesOrNo828) [inline flink/v1.12/flink/src/test/java/com/netease/arctic/flink/table/TestKeyed.java:442]: Could you add a UT when an upsert table functionality is disabled?

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java [hunk 2]
```diff
@@ -213,6 +217,22 @@ public void processElement(StreamRecord<RowData> element) throws Exception {
     });
   }
 
+  /**
+   * If upsert is enabled, turn update_after to insert if there isn't update_after followed by update_before.
+   * But it may lead to some unnecessary delete data if there are some data interspersed with other primary key data.
+   * e.g. K1-UB, K2-UB, K1-UA, K2-UB
+   */
+  private void processMultiUpdateAfter(RowData row) {
```
Comment (YesOrNo828) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java:225]: Maybe we should also consider this scenario:
Key1-UB, Key1-UA, Key2-UB, Key1-UA.

**Review by YesOrNo828 — COMMENTED** [review]
@zstraw Thanks for your contribution, I left some comments.

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java [hunk 3]
```diff
@@ -213,6 +217,22 @@ public void processElement(StreamRecord<RowData> element) throws Exception {
...
+    if (!upsert) {
```
[the remark this answers is not in the document]
Comment (zstraw) [reply flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java:226]: It makes sense to me.

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java (hunk 2 above)
Comment (zstraw) [reply flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java:225]: I will track ua before in keys set to fix this case.

File: flink/v1.12/flink/src/test/java/com/netease/arctic/flink/table/TestKeyed.java (hunk 1 above)
Comment (zstraw) [reply flink/v1.12/flink/src/test/java/com/netease/arctic/flink/table/TestKeyed.java:442]: Ok

**Pull request closed without merging**